Add ```GetEnvironmentVariableValuesAsync``` method to  ResourceExtensions

Alirexaa opened the pull request:
close: #4464 

This change is based on this [EnvironmentVariableEvaluator.GetEnvironmentVariablesAsync](https://github.com/dotnet/aspire/blob/3a974afc395f8f144ebb0cf8a116bd7038970297/tests/Aspire.Hosting.Tests/Utils/EnvironmentVariableEvaluator.cs#L8) method
with little change.

the difference between the issue Proposed API and this commit:

The issue Proposed API :
```
public static class ResourceExtensions
{
   public static Task<IReadOnlyDictionary<string, string?>> GetEnvironmentVariableValuesAsync(this IResourceWithEnvironment resource);
}
```


The commit Proposed API :
```
public static class ResourceExtensions
{
  public static async ValueTask<IReadOnlyDictionary<string, string>> GetEnvironmentVariableValuesAsync(this IResourceWithEnvironment resource,
    DistributedApplicationOperation applicationOperation = DistributedApplicationOperation.Run);
}
```
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4530)

Alirexaa committed:
Add to ResourceExtensions ```GetEnvironmentVariableValuesAsync``` #4464

Alirexaa committed:
Merge branch 'main' into ResourceExtensions

mitchdenny commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -65,6 +65,49 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+    /// </summary>
+    /// <param name="resource">The resource to get the environment variables from.</param>
+    /// <param name="applicationOperation">The context in which the AppHost is being executed</param>
+    /// <returns>The environment variables retrieved from the resource.</returns>
```
Can we get some `<remarks>` added which cover off where this method is useful and considerations around the sync nature. As well as one or two `<example>` sections which cover off its usage.

I'm not sure this needs its own conceptual document, but the following could be expanded:

https://learn.microsoft.com/en-us/dotnet/aspire/fundamentals/testing

Could you create a docs issue to capture that?

mitchdenny commented:
@Alirexaa this is looking good. Just want to make sure we capture the doc requirement and make sure we expand the XML docs a bit for this change.

We need some covering tests as well.

RussKie commented:
/azp run

mitchdenny commented:
@Alirexaa just to be clear we want the XML doc comments in this PR :)

Alirexaa commented:
> @Alirexaa just to be clear we want the XML doc comments in this PR :)

sure :smile:

> Could you create a docs issue to capture that?

@mitchdenny, what did you mean by that?

DamianEdwards commented:
@Alirexaa 

> @mitchdenny, what did you mean by that?

Just that can you please log an issue in dotnet/docs-aspire to ensure the new documentation requirement is captured.

mitchdenny committed:
Merge branch 'main' into ResourceExtensions

Alirexaa committed:
add example and remarks

Alirexaa committed:
add test

DamianEdwards commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -68,11 +68,40 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
     /// </summary>
+    /// <remarks>
+    /// This method is useful when you want to make sure the environment variables are added properly to resources, mostly in test situations.
+    /// This method has asynchronous behavior if <paramref name = "applicationOperation" /> be <see cref="DistributedApplicationOperation.Run"/>
```
```suggestion
    /// This method has asynchronous behavior when <paramref name = "applicationOperation" /> is <see cref="DistributedApplicationOperation.Run"/>
```

DamianEdwards commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -68,11 +68,40 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+    ///      .WithEndpoint("tcp", e => e.AllocatedEndpoint = new AllocatedEndpoint(e, "localhost", 3000, containerHost))
+    ///      .WithMongoExpress();
+    ///
+    ///   var env = await mongoExpress.GetEnvironmentVariableValuesAsync();
```
nit: seems indenting got messed up here

DamianEdwards commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -68,11 +68,40 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
     /// <returns>The environment variables retrieved from the resource.</returns>
+    /// <example>
+    /// <code>
+    /// using var builder = TestDistributedApplicationBuilder.Create();
```
The `TestDistributedApplicationBuilder` type is in this repo only. For the example here you should use `DistributedApplicationTestingBuilder.CreateAsync()`

radical commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -65,6 +65,78 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+                var value = (applicationOperation, expr) switch
+                {
+                    (_, string s) => s,
+                    (DistributedApplicationOperation.Run, IValueProvider provider) => await provider.GetValueAsync().ConfigureAwait(false),
+                    (DistributedApplicationOperation.Publish, IManifestExpressionProvider provider) => provider.ValueExpression,
+                    (_, null) => null,
```
Tests needed for these cases. From code coverage:
<img width="1088" alt="Screenshot 2024-06-27 at 16 00 23" src="https://github.com/dotnet/aspire/assets/1472/1fa07906-c814-4eff-a590-74da1552fcc9">

Alirexaa committed:
Update src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs

Co-authored-by: Damian Edwards <damian@damianedwards.com>

Alirexaa committed:
Update src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

Alirexaa committed:
Update src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

JamesNK commented:
@Alirexaa We have a deadline for 8.1 coming in a couple of weeks. Do you have time to look at the changes? If not, I can help out by addressing some feedback on your branch.

Alirexaa commented:
> @Alirexaa We have a deadline for 8.1 coming in a couple of weeks and we'd like to include this feature. Do you have time to look at the changes soon? If not, I can help out by addressing some feedback on your branch.

Yes, I have time, but I need some examples to write tests for the following cases :
![](https://private-user-images.githubusercontent.com/1472/343936435-1fa07906-c814-4eff-a590-74da1552fcc9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjAwMTY4NzQsIm5iZiI6MTcyMDAxNjU3NCwicGF0aCI6Ii8xNDcyLzM0MzkzNjQzNS0xZmEwNzkwNi1jODE0LTRlZmYtYTU5MC03NGRhMTU1MmZjYzkucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcwMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MDNUMTQyMjU0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MTQ2YTk3OTZlMmE2MzM4ZmI0MjdhYmM5YmZmODU5NmJlMzhkYzgwMDhmZDJiZjAzOGMzZDJiMGU0NWYwYmE3MCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.9XzbLYuB9d3n7wQZix_yvOBBo3N80W4ZqWQHkAPX1AM)

Alirexaa committed:
add more tests,add suggested code

Alirexaa commented:
@JamesNK, please take a look.

mitchdenny commented:
@DamianEdwards @Alirexaa @JamesNK @ReubenBond I've been thinking about this method. At the moment it is in `Aspire.Hosting` but I think given its most common usage scenario it should probably be in `Aspire.Hosting.Testing`.

If we keep it in `Aspire.Hosting` we'd probably want it to extend `IResourceBuilder<T> where T: IResourceWithEnviroment` rather `IResourceBuilderWithEnvironment` directly. Additionally, we'd probably want to remove the operation argument since this would be controlled by the app builder.

@Alirexaa looking at the code coverage on my local run it looks like you figured out how to hit those branches (except the null case and unsupported type case ...)

Alirexaa commented:
> @Alirexaa looking at the code coverage on my local run it looks like you figured out how to hit those branches (except the null case and unsupported type case ...)

Do we need code coverage for null case and unsupported type case?

JamesNK commented:
> @DamianEdwards @Alirexaa @JamesNK @ReubenBond I've been thinking about this method. At the moment it is in `Aspire.Hosting` but I think given its most common usage scenario it should probably be in `Aspire.Hosting.Testing`.

Doesn't this basically already exist in `Aspire.Hosting.Testing`? [EnvironmentVariableEvaluator.GetEnvironmentVariablesAsync](https://github.com/dotnet/aspire/blob/3a974afc395f8f144ebb0cf8a116bd7038970297/tests/Aspire.Hosting.Tests/Utils/EnvironmentVariableEvaluator.cs#L8)

mitchdenny commented:
This is a good point!

mitchdenny commented:
Upon reflection I think that this API is fine the way it is.

Its most likely to be used in something like a lifecycle hook where you don't have access to a builder, but you can resolve the execution context from DI. So, you would pass the operation in.

mitchdenny commented:
/azp run

Alirexaa committed:
Merge branch 'dotnet:main' into ResourceExtensions

eerhardt commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -65,6 +65,78 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+    /// <returns>The environment variables retrieved from the resource.</returns>
+    /// <example>
+    /// <code>
+    /// using var builder = await DistributedApplicationTestingBuilder.CreateAsync&lt;SampleShop_AppHost&gt;();
+    ///
+    /// builder.AddMongoDB("mongo")
+    ///     .WithEndpoint("tcp", e => e.AllocatedEndpoint = new AllocatedEndpoint(e, "localhost", 3000, containerHost))
+    ///     .WithMongoExpress();
+    ///
+    /// var env = await mongoExpress.GetEnvironmentVariableValuesAsync();
```
Does this code build?

1. What is the `mongoExpress` variable? I don't see it defined in the sample.
2. The result of `DistributedApplicationTestingBuilder.CreateAsync` is an `IDistributedApplicationTestingBuilder` which doesn't have an `.AddMongoDB` extension method.

mitchdenny replied:
Yeah, I think a more realistic scenario for this API usage is actually inside a lifecycle hook. It's hard to show that without a much more detailed sample.

eerhardt commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -65,6 +65,78 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+    /// </code>
+    /// </example>
+
+    public static async ValueTask<IReadOnlyDictionary<string, string>> GetEnvironmentVariableValuesAsync(this IResourceWithEnvironment resource,
```
Does this need to return an IReadOnlyDictionary? Can it just be a Dictionary?

Since this is a new object every time the method is called, and no one else references it, what is the point of returning a read-only object? If the caller wants to modify/add/remove from the dictionary, they will need to make a copy of it. This seems more like a `.ToDictionary()` method, which returns a copy.

mitchdenny committed:
Merge branch 'main' into ResourceExtensions

mitchdenny committed:
Merge branch 'ResourceExtensions' of https://github.com/alirexaa/aspire into ResourceExtensions

mitchdenny committed:
Tweak API based on Eric's feedback.

mitchdenny committed:
Use a different example.

eerhardt committed:
Merge remote-tracking branch 'upstream/main' into ResourceExtensions

radical commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -65,6 +65,88 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+/// </code>
+/// </example>
+
+public static async ValueTask<Dictionary<string, string>> GetEnvironmentVariableValuesAsync(this IResourceWithEnvironment resource,
```
```suggestion
    public static async ValueTask<Dictionary<string, string>> GetEnvironmentVariableValuesAsync(this IResourceWithEnvironment resource,
```

radical commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -65,6 +65,88 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+    ///  .WithEnvironment("xpack.security.enabled", "true")
+    ///  .WithEnvironment(context =>
+    ///  {
+    ///      context.EnvironmentVariables["ELASTIC_PASSWORD"] = "123456";
```
Maybe this so credscan does not complain:
```suggestion
    ///      context.EnvironmentVariables["ELASTIC_PASSWORD"] = "p@ssw0rd1";
```

Or we can add this to `./.config/CredScanSuppressions.json`.

radical commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -65,6 +65,88 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+    ///             Assert.Equal("ELASTIC_PASSWORD", env.Key);
+    ///             Assert.False(string.IsNullOrEmpty(env.Value));
+    ///         });
+/// </code>
+/// </example>
```
```suggestion
    /// </code>
    /// </example>
```

eerhardt commented on tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs:
```diff
@@ -25,4 +25,102 @@ public void TryGetContainerImageNameReturnsCorrectFormatWhenShaNotSupplied()
+            env =>
+            {
+                Assert.Equal("{ElasticPassword.value}", env.Value);
+                Assert.False(string.IsNullOrEmpty(env.Value));
```
We should verify the Value is the correct parameter reference in publish mode, not just that it isn't null or empty.

eerhardt replied:
Reactivating. This hasn't been resolved yet.

eerhardt commented on tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs:
```diff
@@ -25,4 +25,102 @@ public void TryGetContainerImageNameReturnsCorrectFormatWhenShaNotSupplied()
+            env =>
+            {
+                Assert.Equal("ELASTIC_PASSWORD", env.Key);
+                Assert.False(string.IsNullOrEmpty(env.Value));
```
Why isn't this

```suggestion
                Assert.Equal("123456", env.Value);
```

?

mitchdenny replied:
Removed password entirely. Its not important what values we assert for the example.

eerhardt replied:
These test comments aren't resolved yet.

radical commented on tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs:
```diff
@@ -25,4 +25,102 @@ public void TryGetContainerImageNameReturnsCorrectFormatWhenShaNotSupplied()
+         .WithEnvironment("xpack.security.enabled", "true")
+         .WithEnvironment(context =>
+         {
+             context.EnvironmentVariables["ELASTIC_PASSWORD"] = "123456";
```
To avoid credscan failure:
```suggestion
             context.EnvironmentVariables["ELASTIC_PASSWORD"] = "p@ssw0rd1";
```

radical commented on tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs:
```diff
@@ -25,4 +25,102 @@ public void TryGetContainerImageNameReturnsCorrectFormatWhenShaNotSupplied()
+            env =>
+            {
+                Assert.Equal("ELASTIC_PASSWORD", env.Key);
+                Assert.False(string.IsNullOrEmpty(env.Value));
```
Should this check for the contents of the string too?

mitchdenny replied:
Fixed.

radical commented on tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs:
```diff
@@ -25,4 +25,102 @@ public void TryGetContainerImageNameReturnsCorrectFormatWhenShaNotSupplied()
+    public async Task GetEnvironmentVariableValuesAsyncReturnCorrectVariablesUsingValueProviderInRunMode()
+    {
+        var builder = DistributedApplication.CreateBuilder();
+        builder.Configuration["Parameters:ElasticPassword"] = "123456";
```
```suggestion
        builder.Configuration["Parameters:ElasticPassword"] = "p@ssw0rd1";
```

radical commented on tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs:
```diff
@@ -25,4 +25,102 @@ public void TryGetContainerImageNameReturnsCorrectFormatWhenShaNotSupplied()
+    public async Task GetEnvironmentVariableValuesAsyncReturnCorrectVariablesUsingManifestExpressionProviderInPublishMode()
+    {
+        var builder = DistributedApplication.CreateBuilder();
+        builder.Configuration["Parameters:ElasticPassword"] = "123456";
```
```suggestion
        builder.Configuration["Parameters:ElasticPassword"] = "p@ssw0rd1";
```

mitchdenny commented on src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs:
```diff
@@ -65,6 +65,88 @@ public static bool TryGetEnvironmentVariables(this IResource resource, [NotNullW
+    ///  .WithEnvironment("xpack.security.enabled", "true")
+    ///  .WithEnvironment(context =>
+    ///  {
+    ///      context.EnvironmentVariables["ELASTIC_PASSWORD"] = "123456";
```
Fair, good point.

mitchdenny committed:
Update src/Aspire.Hosting/ApplicationModel/ResourceExtensions.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

mitchdenny committed:
Remove password from example.

mitchdenny committed:
Indentation fixes.

mitchdenny committed:
Update tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

mitchdenny committed:
Update tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

mitchdenny committed:
PR feedback.

mitchdenny committed:
Namespace fix.

mitchdenny commented:
@joperezr we are going to want to get this one in for 8.1

eerhardt commented on tests/Aspire.Hosting.Tests/ResourceExtensionsTests.cs:
```diff
@@ -25,4 +25,102 @@ public void TryGetContainerImageNameReturnsCorrectFormatWhenShaNotSupplied()
+         .WithEnvironment("xpack.security.enabled", "true")
+         .WithEnvironment(context =>
+         {
+             context.EnvironmentVariables["ELASTIC_PASSWORD"] = "p@ssw0rd1";
```
I think we are still going to get cred scan failures here, but we can try it.

mitchdenny replied:
Will see. @radical was saying that they were in a suppression list.

eerhardt replied:
There are 2 cred scan checks, and my understanding is that only one of them respects the suppression list.

eerhardt reviewed (approved):
LGTM

mitchdenny commented:
/backport to release/8.1
